src/program/snabbvmx/query/query.lua

```diff
@@ -0,0 +1,154 @@
+module(..., package.seeall)
+
+local S = require("syscall")
+local counter = require("core.counter")
+local ffi = require("ffi")
+local lib = require("core.lib")
+local ipv4 = require("lib.protocol.ipv4")
+local ethernet = require("lib.protocol.ethernet")
+local lwaftr = require("apps.lwaftr.lwaftr")
+local lwtypes = require("apps.lwaftr.lwtypes")
+local lwutil = require("apps.lwaftr.lwutil")
+local shm = require("core.shm")
+local top = require("program.top.top")
+
+local select_snabb_instance = top.select_snabb_instance
+local keys = lwutil.keys
+
+local macaddress_t = ffi.typeof[[
+struct { uint8_t ether[6]; }
+]]
+
+-- Get the counter dir from the code.
+local lwaftr_counters_rel_dir = lwaftr.counters_dir
+
+function show_usage (code)
+   print(require("program.lwaftr.query.README_inc"))
+   main.exit(code)
+end
+
+local function sort (t)
+   table.sort(t)
+   return t
+end
+
+local function is_counter_name (name)
+   return lwaftr.counter_names[name] ~= nil
+end
+
+
+function parse_args (raw_args)
+   local handlers = {}
+   function handlers.h() show_usage(0) end
+   local args = lib.dogetopt(raw_args, handlers, "h",
+                             { help="h" })
+   if #args > 0 then show_usage(1) end
+   return nil
+end
+
+local function read_lwaftr_counters (tree)
+   local ret = {}
+   local cnt, cnt_path, value
+   local counters_path = "/" .. tree .. "/" .. lwaftr_counters_rel_dir
+   local counters = shm.children(counters_path)
+   for _, name in ipairs(counters) do
+      cnt_path = counters_path .. name
+      cnt = counter.open(cnt_path, 'readonly')
+      value = tonumber(counter.read(cnt))
+      name = name:gsub(".counter$", "")
+      ret[name] = value
+    end
+   return ret
+end
+
+local function read_apps_counters (tree, app_name)
+   local ret = {}
+   local cnt, cnt_path, value
+   local counters_path = "/" .. tree .. "/apps/" .. app_name .. "/"
+   local counters = shm.children(counters_path)
+   for _, name in ipairs(counters) do
+      cnt_path = counters_path .. name
+      cnt = counter.open(cnt_path, 'readonly')
+      value = tonumber(counter.read(cnt))
+      name = name:gsub(".counter$", "")
+      ret[name] = value
+    end
+   return ret
+end
```

Comment:
To avoid duplicating code I think what it would be best is to make "lwaftr/query.lua:read_lwaftr_counters" public. 

I would make the function more generic and pass a second argument "path", where counters are stored, instead of having counters_path harcoded. I realized there's already a second parameter named `filter` but it's not actually used. The function also returns max_width of counter names, but that output can be ignored.